src/backend/metal/src/command.rs

```diff
@@ -113,128 +259,386 @@ struct CommandBufferInner {
+                pso::Stage::Fragment =>
+                    encoder.set_fragment_buffer(index as _, offset as _, buffer),
+                _ => unimplemented!()
             }
         }
-        for (i, resource) in self.resources_vs.samplers.iter().enumerate() {
-            if let Some(ref sampler) = *resource {
-                encoder.set_vertex_sampler_state(i as _, Some(sampler));
+        Cmd::BindTexture { stage, index, ref texture } => {
+            let texture = texture.as_ref().map(|x| x.as_ref());
+            match stage {
+                pso::Stage::Vertex =>
+                    encoder.set_vertex_texture(index as _, texture),
+                pso::Stage::Fragment =>
+                    encoder.set_fragment_texture(index as _, texture),
+                _ => unimplemented!()
             }
         }
-        // inherit fragment resources
-        for (i, resource) in self.resources_fs.buffers.iter().enumerate() {
-            if let Some((ref buffer, offset)) = *resource {
-                encoder.set_fragment_buffer(i as _, offset as _, Some(buffer));
+        Cmd::BindSampler { stage, index, ref sampler } => {
+            let sampler = sampler.as_ref().map(|x| x.as_ref());
+            match stage {
+                pso::Stage::Vertex =>
+                    encoder.set_vertex_sampler_state(index as _, sampler),
+                pso::Stage::Fragment =>
+                    encoder.set_fragment_sampler_state(index as _, sampler),
+                _ => unimplemented!()
             }
         }
-        for (i, resource) in self.resources_fs.textures.iter().enumerate() {
-            if let Some(ref texture) = *resource {
-                encoder.set_fragment_texture(i as _, Some(texture));
+        Cmd::BindPipeline(ref pipeline_state, ref depth_stencil) => {
+            encoder.set_render_pipeline_state(pipeline_state);
+            if let Some(ref depth_stencil_state) = *depth_stencil {
+                encoder.set_depth_stencil_state(depth_stencil_state);
             }
         }
-        for (i, resource) in self.resources_fs.samplers.iter().enumerate() {
-            if let Some(ref sampler) = *resource {
-                encoder.set_fragment_sampler_state(i as _, Some(sampler));
+        Cmd::Draw { primitive_type, ref vertices, ref instances } => unsafe {
+            msg_send![*encoder,
+                drawPrimitives: primitive_type
+                vertexStart: vertices.start as NSUInteger
+                vertexCount: (vertices.end - vertices.start) as NSUInteger
+                instanceCount: (instances.end - instances.start) as NSUInteger
+                baseInstance: instances.start as NSUInteger
+            ];
+        }
+        Cmd::DrawIndexed { ref index, primitive_type, ref indices, base_vertex, ref instances } => {
+            let index_offset = indices.start as pso::BufferOffset * match index.index_type {
+                MTLIndexType::UInt16 => 2,
```

Comment:
Not a regression, but I guess we should be careful about byte alignment somewhere here ([`indexBufferOffset` must be aligned to 4 bytes](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/1515520-drawindexedprimitives))